src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +602,231 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting a TLS 1.3 record of size
+   *  |record_size|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  GPR_ASSERT(record_size >= SSL3_RT_HEADER_LENGTH + tag_size);
+  *plaintext_size = record_size - SSL3_RT_HEADER_LENGTH - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        const iovec* protected_vec,
+                                        size_t protected_vec_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting the TLS 1.3 payload stored in
+   *  |protected_vec|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  size_t total_vec_length = get_total_length(protected_vec, protected_vec_size);
+  GPR_ASSERT(total_vec_length >= tag_size);
+  *plaintext_size = total_vec_length - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+s2a_decrypt_status s2a_decrypt_payload(
+    s2a_crypter* crypter, iovec& record_header, const iovec* protected_vec,
+    size_t protected_vec_size, iovec& unprotected_vec,
+    size_t* bytes_written, char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  size_t payload_size = get_total_length(protected_vec, protected_vec_size);
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  GPR_ASSERT(tag_size <= payload_size);
+  size_t expected_plaintext_and_record_byte_size = payload_size - tag_size;
+  GPR_ASSERT(expected_plaintext_and_record_byte_size <=
+             SSL3_RT_MAX_PLAIN_LENGTH + /** record byte **/ 1);
+
+  uint8_t sequence[kTlsSequenceSize];
+  sequence_to_bytes(crypter->in_connection, sequence);
+  size_t nonce_size;
+  uint8_t* nonce = s2a_nonce(crypter, sequence, kTlsSequenceSize, &nonce_size);
+  uint8_t* additional_data = s2a_additional_data(
+      sequence, /** sequence size **/ 8, (uint8_t*)record_header.iov_base,
+      record_header.iov_len, payload_size);
+  iovec aad_vec = {(void*)(additional_data + kTlsSequenceSize),
+                   record_header.iov_len};
```

Comment:
Could be just a naming issue, but I was surprised that you skip 8 bytes of additional_data here.

Reply:
Thanks for pointing this out. I had made things overly complicated. For TLS 1.3, the additional data bytes are simply the record header. See https://tools.ietf.org/html/rfc8446#section-5.2.